core/src/main/java/fr/sncf/osrd/speedcontroller/generators/ConstructionAllowanceGenerator.java

```diff
@@ -74,41 +99,68 @@ protected double getFirstGuess() {
             location.updatePosition(schedule.rollingStock.length, step.positionDelta);
         }
         coastingPhase.put(location.getPathPosition(), targetSpeed);
+        var coastingLastPosition = coastingPhase.lastKey();
 
-        // acceleration phase
-        // it is computed in order to get its beginning position
-        // later, this phase will be used in case the user asks for a high margin in a short distance
-        speed = expectedSpeeds.interpolate(endPosition);
-        location = convertPosition(schedule, endPosition);
-        var acceleratingPhase = new SortedDoubleMap();
-        acceleratingPhase.put(location.getPathPosition(), speed);
-        while (speed > coastingPhase.interpolate(location.getPathPosition())
-                && speed > targetSpeed && location.getPathPosition() > initialPosition) {
-            var stepSpeed = speed;
-            var directive = new SpeedDirective(coastingPhase.interpolate(location.getPathPosition()));
-            var step = nextStep(
-                    location,
-                    speed,
-                    schedule.rollingStock,
-                    TIME_STEP,
-                    location.getPathPosition(),
-                    -1,
-                    (integrator) -> integrator.actionToTargetSpeed(directive, schedule.rollingStock, -1));
-            speed = step.finalSpeed;
-            location.updatePosition(schedule.rollingStock.length, step.positionDelta);
-            acceleratingPhase.put(location.getPathPosition(), stepSpeed);
-        }
+        // if the coasting phase does not intersect the accelerating one,
+        // that means there is at least a small interval with a MARECO-like behavior
+        if (coastingLastPosition <= accelerationFirstPosition) {
+            if (coastingPhase.size() > 1) {
+                res.add(new CoastingSpeedController(initialPosition, coastingLastPosition));
+            }
+            var marecoSpeedControllers =
+                    super.getSpeedControllers(schedule, targetSpeed, coastingLastPosition, accelerationFirstPosition);
+            res.addAll(marecoSpeedControllers);
+        } else {
```

Comment:
You can `return res` at the end of the if. So you don't have a big indented `else` block.

Reply:
Done, thank you !